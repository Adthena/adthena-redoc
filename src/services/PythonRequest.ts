import {
  FetchBodyOptions,
  FetchUrlOptions,
  RequestGenerator,
  RequestGeneratorOptions,
} from './RequestGenerator';
import { groupParamsByKey } from '../utils';

const NEW_LINE = '\n';

export class PythonRequest extends RequestGenerator {
  constructor(options: RequestGeneratorOptions) {
    super(options);
  }

  getSourceCode(): string {
    const fetchUrl: FetchUrlOptions = this.buildFetchUrl();
    const fetchHeaders: Headers = this.buildFetchHeaders();
    const fetchOptions: FetchBodyOptions = this.buildFetchBodyOptions();

    return this.generatePythonSyntax(fetchUrl, fetchHeaders, fetchOptions);
  }

  private generatePythonSyntax(
    fetchUrl: FetchUrlOptions,
    fetchHeaders: Headers,
    fetchOptions: FetchBodyOptions,
  ) {
    const imports = `import requests${NEW_LINE}${NEW_LINE}`;
    let pythonUrlLine = 'url =';
    let pythonResponseLine = 'response = requests.request(';
    let pythonHeaders = '';
    let pythonData = '';
    let pythonQueryParams = '';
    let pythonCookies = '';

    if (!fetchUrl.fetchUrl.startsWith('http')) {
      const url = new URL(fetchUrl.fetchUrl, window.location.href);
      pythonUrlLine = `${pythonUrlLine} "${url.href}"`;
    } else {
      pythonUrlLine = `${pythonUrlLine} "${fetchUrl.fetchUrl}"`;
    }

    if (pythonUrlLine) {
      pythonUrlLine = `${pythonUrlLine}${NEW_LINE}${NEW_LINE}`;
    }

    pythonHeaders = JSON.stringify(
      groupParamsByKey(fetchHeaders.entries() as IterableIterator<[string, string]>),
      undefined,
      2,
    );

    if (pythonHeaders && pythonHeaders != '{}') {
      pythonHeaders = `headers = ${pythonHeaders}${NEW_LINE}${NEW_LINE}`;
    }

    if (fetchUrl.queryParams?.toString()) {
      pythonQueryParams = JSON.stringify(
        groupParamsByKey(fetchUrl.queryParams.entries()),
        undefined,
        2,
      );
      pythonQueryParams = `params = ${pythonQueryParams}${NEW_LINE}${NEW_LINE}`;
    }

    if (fetchOptions.body instanceof URLSearchParams) {
      pythonData = `payload = ${fetchOptions.body.toString()} ${NEW_LINE}`;
    } else if (fetchOptions.body) {
      if (fetchOptions.body instanceof Object) {
        try {
          pythonData = `payload = ${JSON.stringify(fetchOptions.body, undefined, 2)}`;
        } catch (err) {
          // Ignore.
        }
      }
      if (!pythonData) {
        pythonData = `payload = '${fetchOptions.body.replace(/'/g, "'\"'\"'")}'${NEW_LINE}`;
      }
    }

    if (pythonData) {
      pythonData = `${pythonData}${NEW_LINE}${NEW_LINE}`;
    }

    if (fetchUrl.cookieParams?.toString()) {
      pythonCookies = `cookies = ${JSON.stringify(
        groupParamsByKey(fetchUrl.cookieParams?.entries()),
        undefined,
        2,
      )}${NEW_LINE}${NEW_LINE}`;
    }

    const params = [
      `"${this.method.toUpperCase()}"`,
      'url',
      pythonData ? 'data=payload' : '',
      pythonHeaders ? 'headers=headers' : '',
      pythonQueryParams ? 'params=params' : '',
      pythonCookies ? 'cookies=cookies' : '',
    ];
    pythonResponseLine += `${params.filter(Boolean).join(', ')})${NEW_LINE}${NEW_LINE}`;

    let logging = `print(response.status_code)${NEW_LINE}`;
    logging += this.getResponseLogging(fetchHeaders.get('accept') || '');

    return [
      imports,
      pythonUrlLine,
      pythonHeaders,
      pythonQueryParams,
      pythonCookies,
      pythonData,
      pythonResponseLine,
      logging,
    ].join('');
  }

  private getResponseLogging(acceptHeader: string) {
    switch (acceptHeader) {
      case 'application/json':
        return 'print(response.json())';
      case 'application/octet-stream':
        return 'print(response.content)';
      case 'application/x-www-form-urlencoded':
      default:
        return 'print(response.text)';
    }
  }
}
