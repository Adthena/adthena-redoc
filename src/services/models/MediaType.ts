import * as Sampler from 'openapi-sampler';

import type { OpenAPIMediaType } from '../../types';
import type { RedocNormalizedOptions } from '../RedocNormalizedOptions';
import { SchemaModel } from './Schema';

import { isXml, mapValues } from '../../utils';
import type { OpenAPIParser } from '../OpenAPIParser';
import { ExampleModel } from './Example';
import { ConfigAccessOptions, FinalExamples, generateXmlExample } from '../../utils/xml';
import { MergedOpenAPISchema } from '../types';
import { OpenAPIExample, Referenced } from '../../types';

export class MediaTypeModel {
  examples?: { [name: string]: ExampleModel };
  schema?: SchemaModel;
  name: string;
  isRequestType: boolean;
  onlyRequiredInSamples: boolean;
  generatedPayloadSamplesMaxDepth: number;

  /**
   * @param isRequestType needed to know if skipe RO/RW fields in objects
   */
  constructor(
    parser: OpenAPIParser,
    name: string,
    isRequestType: boolean,
    info: OpenAPIMediaType,
    options: RedocNormalizedOptions,
  ) {
    this.name = name;
    this.isRequestType = isRequestType;
    this.schema = info.schema && new SchemaModel(parser, info.schema, '', options);
    this.onlyRequiredInSamples = options.onlyRequiredInSamples;
    this.generatedPayloadSamplesMaxDepth = options.generatedPayloadSamplesMaxDepth;
    const isCodeGenerationSupported = options.codeSamplesLanguages.some(lang =>
      name.toLowerCase().includes(lang),
    );
    if (info.examples !== undefined) {
      this.examples = mapValues(
        info.examples,
        example => new ExampleModel(parser, example, name, info.encoding),
      );
    } else if (info.example !== undefined) {
      this.examples = {
        default: new ExampleModel(
          parser,
          { value: parser.deref(info.example).resolved },
          name,
          info.encoding,
        ),
      };
    } else if (isCodeGenerationSupported) {
      this.generateExample(parser, info);
    }
  }

  generateExample(parser: OpenAPIParser, info: OpenAPIMediaType) {
    const samplerOptions = {
      skipReadOnly: this.isRequestType,
      skipWriteOnly: !this.isRequestType,
      skipNonRequired: this.isRequestType && this.onlyRequiredInSamples,
      maxSampleDepth: this.generatedPayloadSamplesMaxDepth,
    };
    if (this.schema) {
      if (this.schema.oneOf) {
        this.examples = {};
        for (const subSchema of this.schema.oneOf) {
          const sample = Sampler.sample(subSchema.rawSchema as any, samplerOptions, parser.spec);

          if (this.schema.discriminatorProp && typeof sample === 'object' && sample) {
            sample[this.schema.discriminatorProp] = subSchema.title;
          }

          this.examples[subSchema.title] = new ExampleModel(
            parser,
            {
              value: sample,
            },
            this.name,
            info.encoding,
          );

          const xmlExamples = this.resolveXmlExample(parser, sample as OpenAPIExample);
          if (xmlExamples[0]) {
            this.examples[subSchema.title].value = xmlExamples[0]?.exampleValue;
          }
        }
      } else {
        const infoOrRef: Referenced<OpenAPIExample> = {
          value: Sampler.sample(info.schema as any, samplerOptions, parser.spec),
        };
        const xmlExamples = this.resolveXmlExample(parser, infoOrRef.value);

        if (xmlExamples.length > 1) {
          this.examples = Object.fromEntries(
            xmlExamples.map(item => [
              item.exampleId,
              new ExampleModel(
                parser,
                {
                  value: item.exampleValue,
                },
                this.name,
                info.encoding,
              ),
            ]),
          );
        } else {
          this.examples = {
            default: new ExampleModel(
              parser,
              {
                value: xmlExamples[0]?.exampleValue || infoOrRef.value,
              },
              this.name,
              info.encoding,
            ),
          };
        }
      }
    }
  }

  resolveXmlExample(parser: OpenAPIParser, sample: OpenAPIExample) {
    const configAccessOptions: ConfigAccessOptions = {
      includeReadOnly: !this.isRequestType,
      includeWriteOnly: this.isRequestType,
    };
    const subSchema = this.schema?.schema;
    let xmlExamples: FinalExamples[] = [];
    if (subSchema && isXml(this.name)) {
      let resolved;
      if (subSchema.items) {
        resolved = {
          ...subSchema,
          items: parser.derefSchemaWithExample(
            subSchema.items as MergedOpenAPISchema,
            Array.isArray(sample) ? sample[0] : sample,
          ),
        };
      } else {
        resolved = parser.derefSchemaWithExample(subSchema, sample);
      }
      xmlExamples = generateXmlExample({
        includeReadOnly: configAccessOptions?.includeReadOnly,
        includeWriteOnly: configAccessOptions?.includeWriteOnly,
        schema: resolved,
      });
    }

    return xmlExamples;
  }
}
