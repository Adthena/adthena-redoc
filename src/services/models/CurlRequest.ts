import { FieldModel } from './Field';
import { SecurityScheme } from './SecurityRequirement';
import { MediaContentModel } from './MediaContent';
import { getSerializedValue } from '../../utils';
import { MediaTypeModel } from './MediaType';

const NEW_LINE = '\\\n';

interface FetchBodyOptions {
  method: string;
  body?: any;
}

/**
 * @license
 * MIT License
 *
 * Copyright (c) 2022 Mrinmoy Majumdar
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
export class CurlRequestModel {
  apiKeys: SecurityScheme[];
  method: string;
  parameters: FieldModel[];
  path: string;
  requestBody?: MediaContentModel;
  serverUrl: string;

  constructor({ apiKeys, method, parameters, path, requestBody, serverUrl }) {
    this.apiKeys = apiKeys;
    this.method = method;
    this.parameters = parameters;
    this.path = path.startsWith('/') ? path : `/${path}`;
    this.requestBody = requestBody;
    this.serverUrl = serverUrl;
  }

  getCurlCommand(): string {
    const fetchUrl = this.buildFetchUrl();
    const fetchHeaders = this.buildFetchHeaders();
    const fetchOptions = this.buildFetchBodyOptions();

    return this.generateCurlSyntax(fetchUrl, fetchHeaders, fetchOptions);
  }

  private buildFetchUrl() {
    let fetchUrl;
    fetchUrl = this.path;
    // Generate URL using Path Params
    this.parameters
      .filter(param => param.in === 'path' && typeof param.example !== 'undefined')
      .forEach(param => {
        fetchUrl = fetchUrl.replace(`{${param.name}}`, encodeURIComponent(param.example as string));
      });

    // Query Params
    const urlQueryParamsMap = new Map();
    const queryParameters = this.parameters.filter(param => param.in === 'query');
    if (queryParameters.length > 0) {
      queryParameters.forEach((param: FieldModel) => {
        const queryParam = new URLSearchParams();
        const value = param.example ?? param.schema.default;

        if (typeof value !== 'undefined') {
          const serializedParam = getSerializedValue(param, value);
          queryParam.append(param.name, serializedParam);
        }
        if (queryParam.toString()) {
          urlQueryParamsMap.set(param.name, queryParam);
        }
      });
    }

    let urlQueryParamString = '';
    if (urlQueryParamsMap.size) {
      urlQueryParamsMap.forEach((val, pname) => {
        urlQueryParamString += `${val.get(pname)}&`;
      });
      urlQueryParamString = urlQueryParamString.slice(0, -1);
    }
    if (urlQueryParamString.length !== 0) {
      fetchUrl = `${fetchUrl}${fetchUrl.includes('?') ? '&' : '?'}${urlQueryParamString}`;
    }

    // Add authentication Query-Param if provided
    this.apiKeys
      .filter(scheme => scheme.in === 'query')
      .forEach(scheme => {
        fetchUrl = `${fetchUrl}${fetchUrl.includes('?') ? '&' : '?'}${
          scheme.name
        }=${encodeURIComponent(scheme.displayName)}`;
      });

    fetchUrl = `${this.serverUrl.replace(/\/$/, '')}${fetchUrl}`;
    return fetchUrl;
  }

  private buildFetchHeaders() {
    const defaultAcceptHeader = 'application/json';
    const reqHeaders = new Headers();

    // Add Authentication Header if provided
    this.apiKeys
      .filter(scheme => scheme.in === 'header')
      .forEach(scheme => {
        reqHeaders.append(<string>scheme.name, scheme.displayName);
      });

    // Add Header Params
    this.parameters
      .filter(param => param.in === 'header')
      .forEach(param => {
        if (param.example) {
          reqHeaders.append(param.name, param.example);
        } else if (Array.isArray(param.examples)) {
          param.examples.forEach(example => {
            reqHeaders.append(param.name, example);
          });
        }
      });

    if (this.requestBody?.mediaTypes.length) {
      const [mediaType] = this.requestBody.mediaTypes;
      if (!mediaType.name.includes('form-data')) {
        // For multipart/form-data dont set the content-type to allow creation of browser generated part boundaries
        reqHeaders.append('Content-Type', mediaType.name);
      }
    }

    if (!reqHeaders.get('Accept')) {
      reqHeaders.append('Accept', defaultAcceptHeader);
    }

    return reqHeaders;
  }

  private buildFetchBodyOptions() {
    const fetchOptions: FetchBodyOptions = {
      method: this.method.toUpperCase(),
    };

    const [mediaType = {} as MediaTypeModel] = this.requestBody?.mediaTypes || [];
    const { examples, name } = mediaType;

    if (examples) {
      if (name.includes('form-urlencoded') && examples) {
        const formUrlParams = new URLSearchParams();
        Object.entries(examples)
          .filter(([, example]) => !example.mime.includes('file'))
          .forEach(([key, example]) => {
            if (!Array.isArray(example.value)) {
              if (example.value) {
                formUrlParams.append(key, example.value);
              }
            } else {
              const vals =
                example.value && Array.isArray(example.value) ? example.value.join(',') : '';
              formUrlParams.append(key, vals);
            }
          });
        fetchOptions.body = formUrlParams;
      } else if (name.includes('form-data')) {
        const formDataParams = new FormData();
        const [{ value }] = Object.values(examples as object);
        const formDataEls = Object.entries(value);
        formDataEls.forEach(([key, propertyValue]) => {
          if (!Array.isArray(propertyValue)) {
            formDataParams.append(key, <string>propertyValue);
          } else if (propertyValue && Array.isArray(propertyValue)) {
            formDataParams.append(key, propertyValue.join(','));
          }
        });
        fetchOptions.body = formDataParams;
      } else if (
        /^audio\/|^image\/|^video\/|^font\/|tar$|zip$|7z$|rtf$|msword$|excel$|\/pdf$|\/octet-stream$/.test(
          name,
        )
      ) {
        const [{ value }] = Object.values(examples as object);
        fetchOptions.body = value;
      } else if (name.includes('json') || name.includes('xml') || name.includes('text')) {
        const [{ value }] = Object.values(examples as object);

        fetchOptions.body = value;
      }
    }

    return fetchOptions;
  }

  private generateCurlSyntax(
    fetchUrl: any,
    fetchHeaders: Headers | any[],
    fetchOptions: FetchBodyOptions,
  ) {
    let curlUrl;
    let curl = '';
    let curlHeaders = '';
    let curlData = '';
    let curlForm = '';

    if (!fetchUrl.startsWith('http')) {
      const url = new URL(fetchUrl, window.location.href);
      curlUrl = url.href;
    } else {
      curlUrl = fetchUrl;
    }

    curl = `curl -X ${this.method.toUpperCase()} "${curlUrl}" ${NEW_LINE}`;

    curlHeaders = [...(fetchHeaders as any[])]
      .map(([key, value]) => ` -H "${key}: ${value}"`)
      .join(` ${NEW_LINE}`);
    if (curlHeaders) {
      curlHeaders = `${curlHeaders} ${NEW_LINE}`;
    }
    if (fetchOptions.body instanceof URLSearchParams) {
      curlData = ` -d ${fetchOptions.body.toString()} ${NEW_LINE}`;
    } else if (fetchOptions.body instanceof File) {
      curlData = ` --data-binary @${fetchOptions.body.name} ${NEW_LINE}`;
    } else if (fetchOptions.body instanceof FormData) {
      curlForm = [...(fetchOptions.body as unknown as any[])]
        .reduce((aggregator, [key, value]) => {
          if (value instanceof File) {
            return [...aggregator, ` -F "${key}=@${value.name}"`];
          }

          const multiple = value.match(/([^,],)/gm);

          if (multiple) {
            const multipleResults = multiple.map(one => `-F "${key}[]=${one}"`);

            return [...aggregator, ...multipleResults];
          }

          return [...aggregator, ` -F "${key}=${value}"`];
        }, [])
        .join(` ${NEW_LINE}`);
    } else if (fetchOptions.body) {
      if (fetchOptions.body instanceof Object) {
        try {
          curlData = ` -d '${JSON.stringify(fetchOptions.body)}' ${NEW_LINE}`;
        } catch (err) {
          // Ignore.
        }
      }
      if (!curlData) {
        curlData = ` -d '${fetchOptions.body.replace(/'/g, "'\"'\"'")}' ${NEW_LINE}`;
      }
    }

    return `${curl}${curlHeaders}${curlData}${curlForm}`;
  }
}
