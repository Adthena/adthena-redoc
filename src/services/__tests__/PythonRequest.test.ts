import { PythonRequest } from '../PythonRequest';
import { FieldModel, MediaContentModel } from '../models';

describe('PythonRequest', () => {
  it('should generate a GET request', () => {
    const request = new PythonRequest({
      apiKeys: [],
      method: 'get',
      parameters: [],
      path: '/pet/1',
      requestBody: undefined,
      serverUrl: 'https://example.com',
    });

    expect(request.getSourceCode()).toMatchSnapshot();
  });

  it('should generate a GET request with parameters', () => {
    const request = new PythonRequest({
      apiKeys: [],
      method: 'get',
      parameters: [
        {
          name: 'petId',
          in: 'path',
          required: true,
          example: 123,
          schema: {
            type: 'integer',
          },
          serializationMime: 'application/json',
        },
        {
          name: 'name',
          in: 'query',
          required: true,
          example: 'Bob',
          schema: {
            type: 'string',
          },
          serializationMime: 'application/json',
        },
        {
          name: 'status',
          in: 'query',
          required: true,
          example: 'available',
          schema: {
            type: 'string',
          },
          serializationMime: 'application/json',
        },
      ] as FieldModel[],
      path: '/pet/{petId}',
      requestBody: undefined,
      serverUrl: 'https://example.com',
    });

    expect(request.getSourceCode()).toMatchSnapshot();
  });

  it('should generate a POST request with parameters', () => {
    const request = new PythonRequest({
      apiKeys: [],
      method: 'post',
      parameters: [
        {
          name: 'petId',
          in: 'path',
          required: true,
          example: 123,
          schema: {
            type: 'integer',
          },
          serializationMime: 'application/json',
        },
        {
          name: 'name',
          in: 'query',
          required: true,
          example: 'Bob',
          schema: {
            type: 'string',
          },
          serializationMime: 'application/json',
        },
        {
          name: 'status',
          in: 'query',
          required: true,
          example: 'available',
          schema: {
            type: 'string',
          },
          serializationMime: 'application/json',
        },
      ] as FieldModel[],
      path: '/pet/{petId}',
      requestBody: {
        mediaTypes: [
          {
            name: 'application/json',
            examples: {
              foo: {
                mime: 'application/json',
                value: {
                  id: '5678',
                  foo: 'bar',
                },
              },
            },
          },
        ],
      } as unknown as MediaContentModel,
      serverUrl: 'https://example.com',
    });

    expect(request.getSourceCode()).toMatchSnapshot();
  });
});
