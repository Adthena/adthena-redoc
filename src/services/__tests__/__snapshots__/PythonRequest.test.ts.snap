// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PythonRequest should generate a GET request 1`] = `
"import requests

url = \\"https://example.com/pet/1\\"

headers = {
  \\"accept\\": \\"application/json\\"
}

response = requests.request(\\"GET\\", url, headers=headers)

print(response.status_code)
print(response.json())"
`;

exports[`PythonRequest should generate a GET request with parameters 1`] = `
"import requests

url = \\"https://example.com/pet/123\\"

headers = {
  \\"accept\\": \\"application/json\\"
}

params = {
  \\"name\\": \\"\\\\\\"Bob\\\\\\"\\",
  \\"status\\": \\"\\\\\\"available\\\\\\"\\"
}

response = requests.request(\\"GET\\", url, headers=headers, params=params)

print(response.status_code)
print(response.json())"
`;

exports[`PythonRequest should generate a POST request with parameters 1`] = `
"import requests

url = \\"https://example.com/pet/123\\"

headers = {
  \\"accept\\": \\"application/json\\",
  \\"content-type\\": \\"application/json\\"
}

params = {
  \\"name\\": \\"\\\\\\"Bob\\\\\\"\\",
  \\"status\\": \\"\\\\\\"available\\\\\\"\\"
}

payload = {
  \\"id\\": \\"5678\\",
  \\"foo\\": \\"bar\\"
}

response = requests.request(\\"POST\\", url, data=payload, headers=headers, params=params)

print(response.status_code)
print(response.json())"
`;
